### Importing libraries

In [1]:
import pandas as pd
import numpy as np

### Loading the dataset

In [3]:
df=pd.read_csv(r'C:\Users\Sanza Vijin\Downloads\student-data.csv')
df.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  internet romantic  famrel  freetime  goout Dalc Walc health absences passed  
0       no       no       4         3      4    1    1      3        6     no  
1      yes       no       5         3      3    1    1      3        4     no  
2      yes       no       4         3      2    2    3      3       10    yes  
3      yes      yes       3         2      2    1    1      5        2    yes  
4       no       no       4         3      2    1    2      5        4    yes  

[5 rows x 31 columns]

In [4]:
df.shape

(395, 31)

In [5]:
df.isna().sum() #finding missing values

school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
passed        0
dtype: int64

### Preprocessing

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 31 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   school      395 non-null    object
 1   sex         395 non-null    object
 2   age         395 non-null    int64 
 3   address     395 non-null    object
 4   famsize     395 non-null    object
 5   Pstatus     395 non-null    object
 6   Medu        395 non-null    int64 
 7   Fedu        395 non-null    int64 
 8   Mjob        395 non-null    object
 9   Fjob        395 non-null    object
 10  reason      395 non-null    object
 11  guardian    395 non-null    object
 12  traveltime  395 non-null    int64 
 13  studytime   395 non-null    int64 
 14  failures    395 non-null    int64 
 15  schoolsup   395 non-null    object
 16  famsup      395 non-null    object
 17  paid        395 non-null    object
 18  activities  395 non-null    object
 19  nursery     395 non-null    object
 20  higher    

In [7]:
x = df.drop(['age','address','famsize','Pstatus','Medu','Fedu','reason','guardian','famsup','paid','nursery','higher','internet',
              'romantic','famrel','schoolsup','passed'],axis=1)
y = pd.DataFrame(df['passed'])

In [8]:
df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'passed'],
      dtype='object')

In [9]:
x['school'].value_counts()

GP    349
MS     46
Name: school, dtype: int64

In [10]:
df['activities'].value_counts()

yes    201
no     194
Name: activities, dtype: int64

In [11]:
x=pd.get_dummies(x,columns=['activities','sex','school']) ##onehot encoding

In [12]:
y=pd.get_dummies(y,columns=['passed'])

In [13]:
x.head()

Mjob      Fjob  traveltime  studytime  failures  freetime  goout  Dalc  \
0  at_home   teacher           2          2         0         3      4     1   
1  at_home     other           1          2         0         3      3     1   
2  at_home     other           1          2         3         3      2     2   
3   health  services           1          3         0         2      2     1   
4    other     other           1          2         0         3      2     1   

   Walc  health  absences  activities_no  activities_yes  sex_F  sex_M  \
0     1       3         6              1               0      1      0   
1     1       3         4              1               0      1      0   
2     3       3        10              1               0      1      0   
3     1       5         2              0               1      1      0   
4     2       5         4              1               0      1      0   

   school_GP  school_MS  
0          1          0  
1          1          0  
2          1          0  
3          1          0  
4          1          0

In [14]:
x.shape

(395, 17)

In [15]:
x.columns

Index(['Mjob', 'Fjob', 'traveltime', 'studytime', 'failures', 'freetime',
       'goout', 'Dalc', 'Walc', 'health', 'absences', 'activities_no',
       'activities_yes', 'sex_F', 'sex_M', 'school_GP', 'school_MS'],
      dtype='object')

### feature engineering

In [17]:

x['pjob']=x['Mjob']+['Fjob']
x['freetimes']=x['freetime']+x['goout']
x=x.drop(['Mjob','Fjob','freetime','goout'],axis=1)

In [18]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   traveltime      395 non-null    int64 
 1   studytime       395 non-null    int64 
 2   failures        395 non-null    int64 
 3   Dalc            395 non-null    int64 
 4   Walc            395 non-null    int64 
 5   health          395 non-null    int64 
 6   absences        395 non-null    int64 
 7   activities_no   395 non-null    uint8 
 8   activities_yes  395 non-null    uint8 
 9   sex_F           395 non-null    uint8 
 10  sex_M           395 non-null    uint8 
 11  school_GP       395 non-null    uint8 
 12  school_MS       395 non-null    uint8 
 13  pjob            395 non-null    object
 14  freetimes       395 non-null    int64 
dtypes: int64(8), object(1), uint8(6)
memory usage: 30.2+ KB


In [19]:
x=x.drop(['pjob'],axis=1)

#splitting the data

In [20]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.3,random_state=42)

### Building the model

In [21]:
from sklearn.ensemble import RandomForestClassifier
rf=RandomForestClassifier()
rf.fit(x_train,y_train)

y_pred = rf.predict(x_test)

In [22]:
from sklearn.metrics import f1_score,confusion_matrix
print('f1_score:',f1_score(y_test,y_pred,average='weighted'))
print('confusion_matrix:',confusion_matrix(y_test.values.argmax(axis=1),y_pred.argmax(axis=1)))

f1_score: 0.6007651943042492
confusion_matrix: [[11 35]
 [ 5 68]]


In [23]:

#feature importance
pd.Series(rf.feature_importances_,index=x.columns).sort_values(ascending = False)*100

absences          18.428545
freetimes         16.856201
failures          11.498190
health            10.947242
Walc               8.921693
studytime          7.800767
traveltime         5.916803
Dalc               5.913066
sex_M              2.854763
activities_yes     2.743986
activities_no      2.645499
sex_F              2.421423
school_GP          1.576639
school_MS          1.475184
dtype: float64

In [24]:
rf.fit(x_train,y_train)

RandomForestClassifier()

In [25]:
rft= RandomForestClassifier(n_estimators=500,max_depth=10,random_state=42,criterion="entropy")
rft.fit(x_train,y_train)

RandomForestClassifier(criterion='entropy', max_depth=10, n_estimators=500,
                       random_state=42)

In [26]:
y_pred=rft.predict(x_test)
f1_score(y_test,y_pred,average='weighted')

0.5908674614623806

The possibility of a student to pass is 59% according to this model